node/core/dispute-coordinator/src/initialized.rs

```diff
@@ -341,6 +348,147 @@ impl Initialized {
 		Ok(())
 	}
 
+	/// For each unapplied (past-session) slash, report an unsigned extrinsic
+	/// to the runtime.
+	async fn process_unapplied_slashes<Context>(
+		&mut self,
+		ctx: &mut Context,
+		relay_parent: Hash,
+		unapplied_slashes: Vec<(SessionIndex, CandidateHash, vstaging::slashing::PendingSlashes)>,
+	) {
+		for (session_index, candidate_hash, pending) in unapplied_slashes {
+			gum::info!(
+				target: LOG_TARGET,
+				?session_index,
+				?candidate_hash,
+				n_slashes = pending.keys.len(),
+				"Processing unapplied validator slashes",
+			);
+
+			let inclusions = self.scraper.get_blocks_including_candidate(&candidate_hash);
+			debug_assert!(
+				!inclusions.is_empty(),
+				"Couldn't find inclusion parent for an unapplied slash",
+			);
+
+			// Find the first inclusion parent that we can use
+			// to generate key ownership proof on.
+			// We use inclusion parents because of the proper session index.
+			let mut key_ownership_proofs = Vec::new();
+			let mut dispute_proofs = Vec::new();
+
+			for (_height, inclusion_parent) in inclusions {
+				for (validator_index, validator_id) in pending.keys.iter() {
+					let res =
+						key_ownership_proof(ctx.sender(), inclusion_parent, validator_id.clone())
```

Comment:
I'm curious what's the purpose of 'key ownership proof' in general? To proof that a specific session key (which was used to sign the dispute votes?) belongs to a specific account key (holding funds) so that we know who to slash?

Reply:
To prove validator's staking information (e.g. the nominated amount along with who nominated and how much) in the past. We don't store that onchain for old sessions as the amount would be massive. Instead, `Historical` pallet stores the merkle root of this data, which allows runtime to verify the proofs on-chain, while the proofs are generated off-chain (by querying the state of an old block).